src/util/psbt/mod.rs

```diff
@@ -88,6 +91,82 @@ impl PartiallySignedTransaction {
 
         Ok(())
     }
+
+    /// Decide whether the current Input spends segwit outputs or not and
+    /// calculate the corresponding Sighash.
+    /// #Panics:
+    /// Panics if the index >= number of inputs in psbt
+    pub fn signature_hash(&self, idx: usize) -> Result<SigHash, self::Error> {
+        let inp = &self.inputs[idx];
+        let sighash_type = inp.sighash_type.unwrap_or_else(|| SigHashType::All);
+        // Compute Script code for the Script. When this script is used as scriptPubkey, 
+        // the Script Code determines what goes into sighash.
+        // 1) For non-p2sh non-segwit outputs, script code is the public key.
+        let sighash = if let Some(ref non_witness_utxo) = inp.non_witness_utxo {
```

Comment:
would

```rust
match (inp.non_witness_utxo, inp.witness_utxo) {
    (Some(non_witness_utxo), None) => { ... }
    (None, _) => { ... }
    (Some(_), Some(_)) => { /* error 2 */ }
    _ => { /* error 2 */ }
}
```
work better since it will cover currently uncovered situation if both non-witness and witness UTXO are present for the same output?

Reply:
Agreed, it would be good to catch this case and error out.

Reply:
@apoelstra @dr-orlovsky , I tried to follow the specification line by line write https://github.com/bitcoin/bips/blob/master/bip-0174.mediawiki#simple-signer-algorithm. I think the specification does not error if both are present and this is the expected behaviour

Reply:
In fact, from the BIP. An input can have both PSBT_IN_NON_WITNESS_UTXO and PSBT_IN_WITNESS_UTXO.

Reply:
Ok, but which of them has priority over the other?

Reply:
Yes, I followed the algorithm listed https://github.com/bitcoin/bips/blob/master/bip-0174.mediawiki#simple-signer-algorithm.

Reply:
Ok, at least this is according to the spec... However it's wired, will look into spec to see if there could be an issue with it...